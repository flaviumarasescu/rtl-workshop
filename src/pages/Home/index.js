import React from 'react'
import useCheckLoggedIn from "../../hooks/useCheckLoggedIn";

const Home = () => {

    const isLoggedIn = useCheckLoggedIn()

    return <div>
        <h1>Home</h1>
        <p>Logged in: {isLoggedIn.toString()}</p>
    </div>
}

export default Home