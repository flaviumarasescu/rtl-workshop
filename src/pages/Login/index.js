import React, {useEffect, useState} from 'react'
import {useNavigate} from "react-router-dom";
import {useForm} from "react-hook-form";
import {yupResolver} from "@hookform/resolvers/yup";
import * as yup from "yup"

import Form from "../../components/Form";

import './style.css'

const schema = yup
    .object({
        email: yup.string().email('Enter a valid email').required(),
        password: yup.string().min(3, 'Password must contain at least 3 characters').required(),
    })
    .required()

const Login = () => {
    const navigate = useNavigate()

    const [loading, setLoading] = useState(true)

    const {
        register,
        handleSubmit,
        formState: { errors },
        setValue
    } = useForm({resolver: yupResolver(schema)})
    const onSubmit = (data) => {
        localStorage.setItem('userEmail', data.email);
        navigate('/')
    }

    useEffect(() => {
        const fetchUserData = async () => {
            try {
                const response = await fetch('http://localhost:3001/users/1');
                const userData = await response.json();
                setValue('email', userData.email);
                setValue('password', userData.password);
            } catch (error) {
                console.error('Error fetching user data:', error);
            }
            setLoading(false)
        };

        fetchUserData();
    }, []);

    return <>
    {
        loading ? <div data-testid='spinner'></div>
            :
        <div>
            <h1>Login</h1>
            <Form handleSubmit={handleSubmit(onSubmit)}>
                <label htmlFor='email'>Email</label>
                <input id='email' placeholder='email' {...register("email")} type='email' autoComplete='off'/>
                {errors.email?.message}

                <label htmlFor='password'>Password</label>
                <input id='password' placeholder='password' {...register("password")} type='password'
                       autoComplete='new-password'/>
                {errors.password?.message}

                <input type="submit"/>
            </Form>
        </div>
    }
    </>
}

export default Login