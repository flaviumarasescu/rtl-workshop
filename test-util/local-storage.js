export const getLocalStorage = (id) => {
    window.localStorage.getItem(id)
}

export const setLocalStorage = (id, data) => {
    window.localStorage.setItem(id, JSON.stringify(data))
}