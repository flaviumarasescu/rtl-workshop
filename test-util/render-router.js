import React, {cloneElement, isValidElement} from 'react'
import {render, waitFor,screen} from '@testing-library/react'
import { RouterProvider, createMemoryRouter } from 'react-router'

export const renderWithRouter = (children, routes = [], ui, provider) => {
    const options = isValidElement(children)
        ? { element: children, path: '/'}
        : children;

    const router = createMemoryRouter([{ ...options}, ...routes], {
        initialEntries: [options.path],
        initialIndex: 1
    })

    const wrappedUi = provider ? cloneElement(provider, {}, ui) : ui

    const finishLoading = () => waitFor(() => expect(screen.queryByTestId('spinner')).not.toBeInTheDocument())

    const utils = render(<RouterProvider router={router}>
        {wrappedUi}
    </RouterProvider>)

    return {
        ...utils,
        finishLoading
    }
}