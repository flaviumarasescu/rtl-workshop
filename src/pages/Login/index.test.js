import React from 'react';
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import nock from 'nock';
import { renderWithRouter } from "../../../test-util/render-router";
import Login from "./index";
import Home from "../Home";

nock('http://localhost:3001')
    .defaultReplyHeaders({
    'access-control-allow-origin': '*',
    'access-control-allow-credentials': 'true'
    })
    .persist()
    .get('/users/1')
    .reply(200, { email: 'test1', password: 'p' });

console.log('Nock interceptors:', nock.activeMocks());

describe('Login page', ()=>{
    it('should render the form', async () => {
        const { finishLoading } = renderWithRouter(<Login/>)

        await finishLoading()

        expect(screen.getByRole('heading', { level: 1, name: 'Login' })).toBeInTheDocument()
        expect(screen.getByLabelText(/email/i).value).toBe('test1')
        expect(screen.getByPlaceholderText(/password/i).value).toBe('p')
        expect(screen.getByRole('button', { name: /submit/i})).toBeInTheDocument()
    })

    it('should validate form fields', async () => {
        const { finishLoading } = renderWithRouter(<Login/>)

        await finishLoading()

        const user = userEvent.setup()
        await user.type(screen.getByLabelText('Email'), 'test')
        await user.type(screen.getByLabelText('Password'), 'a')
        await user.click(screen.getByRole('button'))

        expect(screen.getByText(/enter a valid email/i)).toBeInTheDocument()
        expect(screen.getByText(/password must contain at least 3 characters/i)).toBeInTheDocument()

        await user.type(screen.getByLabelText('Email'), 'test@test.com')
        await user.type(screen.getByLabelText('Password'), 'password')
        await user.click(screen.getByRole('button'))

        expect(screen.queryByText(/enter a valid email/i)).not.toBeInTheDocument()
        expect(screen.queryByText(/password must contain at least 3 characters/i)).not.toBeInTheDocument()
    })

    test('form completion and redirect', async () => {
        const { finishLoading } = renderWithRouter({
        element: <Login/>, path:'/login'
    }, [{
            element: <Home/>,
            path: '/'
        }])

        await finishLoading()

        const user = userEvent.setup()
        await user.type(screen.getByLabelText('Email'), 'test@test.com')
        await user.type(screen.getByLabelText('Password'), 'test@test.com')
        await user.click(screen.getByRole('button'))

        expect(screen.getByRole('heading', { level: 1 })).toBeInTheDocument()

    })

    it('renders correctly', async () => {
        const { finishLoading, container } = renderWithRouter(<Login/>)

        await finishLoading()

        expect(container.firstChild).toMatchSnapshot()
    })
})