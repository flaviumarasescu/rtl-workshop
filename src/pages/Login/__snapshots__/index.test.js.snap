// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page renders correctly 1`] = `
<div>
  <h1>
    Login
  </h1>
  <form
    autocomplete="off"
  >
    <label
      for="email"
    >
      Email
    </label>
    <input
      autocomplete="off"
      id="email"
      name="email"
      placeholder="email"
      type="email"
    />
    <label
      for="password"
    >
      Password
    </label>
    <input
      autocomplete="new-password"
      id="password"
      name="password"
      placeholder="password"
      type="password"
    />
    <input
      type="submit"
    />
  </form>
</div>
`;
